kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt

```diff
@@ -0,0 +1,616 @@
+    }
+    is TypeVariableName -> TODO()
+  // Shouldn't happen
+    else -> throw IllegalArgumentException("Unrepresentable type: " + this)
+  }
+}
+
+private data class Property(
+    val name: String,
+    val fqClassName: String,
+    val serializedName: String,
+    val hasDefault: Boolean,
+    val nullable: Boolean,
+    val typeName: TypeName,
+    val jsonQualifiers: List<AnnotationMirror>)
+
+private data class Adapter(
+    val fqClassName: String,
+    val packageName: String,
+    val propertyList: List<Property>,
+    val originalElement: Element,
+    val name: String = fqClassName.substringAfter(packageName)
+        .replace('.', '_')
+        .removePrefix("_"),
+    val hasCompanionObject: Boolean,
+    val visibility: Visibility,
+    val elementUtils: Elements) {
+  fun generate(adapterName: String, fileSpecBuilder: FileSpec.Builder) {
+    val nameAllocator = NameAllocator()
+    fun String.allocate() = nameAllocator.newName(this)
+
+    val originalTypeName = originalElement.asType().asTypeName() as ClassName
+    val moshiName = "moshi".allocate()
+    val moshiParam = ParameterSpec.builder(moshiName, Moshi::class.asClassName()).build()
+    val reader = ParameterSpec.builder("reader".allocate(),
+        JsonReader::class.asClassName()).build()
+    val writer = ParameterSpec.builder("writer".allocate(),
+        JsonWriter::class.asClassName()).build()
+    val value = ParameterSpec.builder("value".allocate(),
+        originalTypeName.asNullable()).build()
+    val jsonAdapterTypeName = ParameterizedTypeName.get(JsonAdapter::class.asClassName(),
+        originalTypeName)
+
+    // Create fields
+    val adapterProperties = propertyList
+        .map { it.typeName.asNonNullable() }
+        .distinct()
+        .associate { typeName ->
+          val propertyName = "${typeName.simplifiedName().allocate()}_Adapter"
+          val adapterTypeName = ParameterizedTypeName.get(JsonAdapter::class.asTypeName(), typeName)
+          typeName to PropertySpec.builder(propertyName, adapterTypeName, PRIVATE)
+              .initializer("%N.adapter%L(%L)",
+                  moshiParam,
+                  if (typeName is ClassName) "" else CodeBlock.of("<%T>", typeName),
+                  typeName.makeType(elementUtils))
+              .build()
+        }
+
+    // TODO in the future make these propertyspecs directly.
+    // Pending https://github.com/square/kotlinpoet/pull/317
+    val allocatedNames = propertyList.associate { it to it.name.allocate() }
+    val optionsByIndex = propertyList
+        .associateBy { it.serializedName }.entries.withIndex()
+
+    // selectName() API setup
+    val optionsCN = JsonReader.Options::class.asTypeName()
+    val optionsProperty = PropertySpec.builder(
+        "OPTIONS".allocate(),
+        optionsCN,
+        PRIVATE)
+        .delegate(
+            "lazy { %T.of(${optionsByIndex.map { it.value.key }
```

Comment:
Are you sure the not very costly operation of creating some constants is worth an anonymous inner class for each MoshiSerializable data class adapter?

Reply:
I'm not! But creating options does quite a bit under the hood, and was hoping to gather feedback here

Reply:
Classloading already makes this lazy. The explicitly lazy isn't needed.

Reply:
aa0534e